src/utils/settings.rb

```diff
@@ -0,0 +1,920 @@
+            gh = group["headers"]
+            if gh
+                headers.concat gh
+            end
+        end
+        headers.each do |h|
+            if h
+                buf << "#include \"#{h}\"\n"
+            end
+        end
+        buf << "\n"
+        buf << prog.string
+        mktmpdir do |dir|
+            file = File.join(dir, "test.cpp")
+            File.open(file, 'w') {|file| file.write(buf.string)}
+            dputs "Compiling #{buf.string}"
+            stdout, stderr = @compiler.run(file, File.join(dir, "test"))
+            dputs "Output: #{stderr}"
+            stderr
+        end
+    end
+
+    def check_conditions
+        buf = StringIO.new
+        conditions = Set.new
+        add_condition = -> (c) {
+            if c && !conditions.include?(c)
+                conditions.add(c)
+                buf << "#if #{c}\n"
+                buf << "#pragma message(#{c.inspect})\n"
+                buf << "#endif\n"
+            end
+        }
+
+        foreach_member do |group, member|
+            add_condition.call(group["condition"])
+            add_condition.call(member["condition"])
+        end
+
+        stderr = compile_test_file(buf)
+        @true_conditions = Set.new
+        stderr.scan(/#pragma message\(\"(.*)\"\)/).each do |m|
+            @true_conditions << m[0]
+        end
+    end
+
+    def sanitize_fields(all=false)
+        pending_types = Hash.new
+        has_booleans = false
+
+        block  = Proc.new do |group, member|
+            if !group["name"]
+                raise "Missing group name"
+            end
+            if !member["name"]
+                raise "Missing member name in group #{group["name"]}"
+            end
+            table = member["table"]
+            if table
+                if !@tables[table]
+                    raise "Member #{member["name"]} references non-existing table #{table}"
+                end
+                @used_tables << table
+            end
+            if !member["field"]
+                member["field"] = member["name"]
+            end
+            typ = member["type"]
+            if !typ
+                pending_types[member] = group
+            elsif typ == "bool"
+                has_booleans = true
+                member["type"] = "uint8_t"
+                member["table"] = OFF_ON_TABLE["name"]
+            end
+        end
+
+        all ? foreach_member(&block) : foreach_enabled_member(&block)
+
+        if has_booleans
+            @tables[OFF_ON_TABLE["name"]] = OFF_ON_TABLE
+            @used_tables << OFF_ON_TABLE["name"]
+        end
+
+        resolve_types pending_types unless !pending_types
+        foreach_enabled_member do |group, member|
+            @count += 1
+            @max_name_length = [@max_name_length, member["name"].length].max
+            if member["table"]
+                @enabled_tables << member["table"]
+            end
+        end
+    end
+
+    def resolve_types(pending)
```

Comment:
IMO this is far too fragile. `settings.yaml` should (eventually) contain enough data to infer types ...

Reply:
Yes, completely agree.

I may have a nice way round this using Clang's libtooling, but haven't had a chance to test this yet.